crates/interpreter/src/interpreter/shared_memory.rs

```diff
@@ -85,53 +84,44 @@ impl SharedMemory {
     #[cfg(feature = "memory_limit")]
     #[inline]
     pub fn limit_reached(&self, new_size: usize) -> bool {
-        (self.last_checkpoint() + new_size) as u64 > self.memory_limit
+        (self.last_checkpoint + new_size) as u64 > self.memory_limit
     }
 
     /// Prepares the shared memory for a new context.
     #[inline]
-    pub fn new_context_memory(&mut self) {
-        let base_offset = self.last_checkpoint();
-        let new_checkpoint = base_offset + self.current_len;
-
+    pub fn new_context(&mut self) {
+        let new_checkpoint = self.buffer.len();
         self.checkpoints.push(new_checkpoint);
-        self.current_len = 0;
+        self.last_checkpoint = new_checkpoint;
     }
 
     /// Prepares the shared memory for returning to the previous context.
     #[inline]
-    pub fn free_context_memory(&mut self) {
+    pub fn free_context(&mut self) {
         if let Some(old_checkpoint) = self.checkpoints.pop() {
-            let last_checkpoint = self.last_checkpoint();
-            self.current_len = old_checkpoint - last_checkpoint;
+            self.last_checkpoint = self.checkpoints.last().cloned().unwrap_or_default();
+            // safety: buffer len is less than or equal `old_checkpoint`
+            unsafe { self.buffer.set_len(old_checkpoint) };
         }
     }
 
     /// Returns the length of the current memory range.
     #[inline]
     pub fn len(&self) -> usize {
-        self.current_len
+        self.buffer.len() - self.last_checkpoint
     }
 
     /// Returns `true` if the current memory range is empty.
     #[inline]
     pub fn is_empty(&self) -> bool {
-        self.current_len == 0
+        self.len() == 0
     }
 
     /// Resizes the memory in-place so that `len` is equal to `new_len`.
     #[inline]
-    pub fn resize(&mut self, new_len: usize) {
-        let last_checkpoint = self.last_checkpoint();
-        let range = last_checkpoint + self.current_len..last_checkpoint + new_len;
-
-        if let Some(available_memory) = self.data.get_mut(range) {
-            available_memory.fill(0);
-        } else {
-            self.data.resize(last_checkpoint + new_len, 0);
-        }
-
-        self.current_len = new_len;
+    pub fn resize(&mut self, new_size: usize) {
+        self.buffer.reserve(usize::max(new_size, 4 * 1024));
```

Comment:
`resize` allocates this, actually for `new_size > 4*1024` it would allocate twice.
Do you mean `usize::min`?

Reply:
My idea was to reserve a minimum of 4096 bytes to avoid many mini-allocations, but I agree that if `new_size > 4 * 1024` then this line is useless

Reply:
Yes, then it should be `min` not `max`

Reply:
Hmm I'm not sure: suppose `new_size` equals 32, if we use `min` then:
```rust
assert_eq!(usize::min(32, 4*1024), 32);
```
and we reserve 32 more bytes instead of 4096, which is not what we want.
I don't understand if I'm missing something here from what you said

Reply:
Ah you're right. This should be removed because `Vec` already does exponential allocations so this is redundant since we're starting already with 4*1024

Reply:
To be precise it should be something like `. reserve(self.last_checkpoint+max(new_size,4*1024))` to allocate at least 4048, but as @DaniPopes said `resize` doubles capacity every time so it is good to rely on.